subsys/mgmt/serial_util.c

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright Runtime.io 2018. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <assert.h>
+#include <stdbool.h>
+#include <crc16.h>
+#include <misc/byteorder.h>
+#include "mgmt/serial.h"
+#include "base64/base64.h"
+
+static u16_t mcumgr_serial_calc_crc(const u8_t *data, int len)
+{
+	return crc16(data, len, 0x1021, 0, true);
+}
+
+/**
+ * Base64-decodes an mcumgr request.  The source and destination buffers can
+ * overlap.
+ */
+static int mcumgr_serial_decode_req(const u8_t *src, u8_t *dst,
+				    int max_dst_len)
+{
+	if (base64_decode_len((char *)src) > max_dst_len) {
+		return -1;
+	}
+
+	return base64_decode(src, dst);
+}
+
+static int mcumgr_serial_parse_op(const u8_t *buf, int len)
+{
+	u16_t op;
+
+	if (len < sizeof(op)) {
+		return -1;
+	}
+
+	memcpy(&op, buf, sizeof(op));
+	op = sys_be16_to_cpu(op);
+
+	if (op != MCUMGR_SERIAL_HDR_PKT && op != MCUMGR_SERIAL_HDR_FRAG) {
+		return -1;
+	}
+
+	return op;
+}
+
+static int mcumgr_serial_parse_len(struct mcumgr_serial_rx_ctxt *rx_ctxt)
+{
+	u16_t len;
+
+	if (rx_ctxt->raw_off < sizeof(len)) {
+		return -1;
+	}
+
+	memcpy(&len, rx_ctxt->buf, sizeof(len));
+	rx_ctxt->pkt_len = sys_be16_to_cpu(len);
+	return rx_ctxt->pkt_len;
+}
+
+static int mcumgr_serial_decode_frag(struct mcumgr_serial_rx_ctxt *rx_ctxt,
+				     const u8_t *frag)
+{
+	int dec_len;
+
+	dec_len = mcumgr_serial_decode_req(
+		frag,
+		rx_ctxt->buf + rx_ctxt->raw_off,
+		rx_ctxt->buf_size - rx_ctxt->raw_off);
+	if (dec_len == -1) {
+		return -1;
+	}
+
+	rx_ctxt->raw_off += dec_len;
+
+	return 0;
+}
+
+/**
+ * Processes a received mcumgr frame.
+ *
+ * @return                      true if a complete packet was received;
+ *                              false if the frame is invalid or if additional
+ *                                  fragments are expected.
+ */
+static bool mcumgr_serial_process_frag(struct mcumgr_serial_rx_ctxt *rx_ctxt,
+				       const u8_t *frag, int frag_len)
+{
+	u16_t crc;
+	u16_t op;
+	int rc;
+
+	op = mcumgr_serial_parse_op(frag, frag_len);
+	switch (op) {
+	case MCUMGR_SERIAL_HDR_PKT:
+		rx_ctxt->raw_off = 0;
+		break;
+
+	case MCUMGR_SERIAL_HDR_FRAG:
+		if (rx_ctxt->raw_off == 0) {
+			return false;
+		}
+		break;
+
+	default:
+		return false;
+	}
+
+	rc = mcumgr_serial_decode_frag(rx_ctxt, frag + sizeof(op));
+	if (rc != 0) {
+		return false;
+	}
+
+	if (op == MCUMGR_SERIAL_HDR_PKT) {
+		rc = mcumgr_serial_parse_len(rx_ctxt);
+		if (rc == -1) {
+			return false;
+		}
+	}
+
+	if (rx_ctxt->raw_off > rx_ctxt->pkt_len + 2) {
```

Comment:
if (rx_ctxt->raw_off != rx_ctxt->pkt_len + 2) {
  return false;
}

Reply:
Huh... not sure what I was thinking.  Anyway, it is fixed now.